ecosystem/python/sdk/aptos_sdk/bcs.py

```diff
@@ -0,0 +1,341 @@
+    ) -> Dict[typing.Any, typing.Any]:
+        length = self.uleb128()
+        values = {}
+        while len(values) < length:
+            key = key_decoder(self)
+            value = value_decoder(self)
+            values[key] = value
+        return values
+
+    def sequence(
+        self,
+        value_decoder: typing.Callable[[Deserializer], typing.Any],
+    ) -> List[typing.Any]:
+        length = self.uleb128()
+        values = []
+        while len(values) < length:
+            values.append(value_decoder(self))
+        return values
+
+    def str(self) -> str:
+        return self.bytes().decode()
+
+    def struct(self, struct: typing.Any) -> typing.Any:
+        return struct.deserialize(self)
+
+    def u8(self) -> int:
+        return self._read_int(1)
+
+    def u16(self) -> int:
+        return self._read_int(2)
+
+    def u32(self) -> int:
+        return self._read_int(4)
+
+    def u64(self) -> int:
+        return self._read_int(8)
+
+    def uleb128(self) -> int:
+        value = 0
+        shift = 0
+
+        while value <= MAX_U128:
+            byte = self._read_int(1)
+            value |= (byte & 0x7F) << shift
+            if byte & 0x80 == 0:
+                break
+            shift += 7
+
+        if value > MAX_U128:
+            raise Exception("Unexpectedly large uleb128 value")
+
+        return value
+
+    def _read(self, length: int) -> bytes:
+        value = self._input.read(length)
+        if value is None or len(value) < length:
+            actual_length = 0 if value is None else len(value)
+            error = (
+                f"Unexpected end of input. Requested: {length}, found: {actual_length}"
+            )
+            raise Exception(error)
+        return value
+
+    def _read_int(self, length: int) -> int:
+        return int.from_bytes(self._read(length), byteorder="little", signed=False)
+
+
+class Serializer:
+    _output: io.BytesIO
+
+    def __init__(self):
+        self._output = io.BytesIO()
+
+    def output(self) -> bytes:
+        return self._output.getvalue()
+
+    def bool(self, value: bool):
+        self._write_int(int(value), 1)
+
+    def bytes(self, value: bytes):
+        self.uleb128(len(value))
+        self._output.write(value)
+
+    def fixed_bytes(self, value):
+        self._output.write(value)
+
+    def map(
+        self,
+        values: typing.Dict[typing.Any, typing.Any],
+        key_encoder: typing.Callable[[Serializer, typing.Any], bytes],
+        value_encoder: typing.Callable[[Serializer, typing.Any], bytes],
+    ):
+        encoded_values = []
+        for (key, value) in values.items():
+            encoded_values.append(
+                (encoder(key, key_encoder), encoder(value, value_encoder))
+            )
+        encoded_values.sort(key=lambda item: item[0])
+
+        self.uleb128(len(encoded_values))
+        for (key, value) in encoded_values:
+            self.fixed_bytes(key)
+            self.fixed_bytes(value)
```

Comment:
Should you sort the map by the keys. Otherwise the iteration order is undetermined.

Reply:
Python3.6 and earlier the dict iteration is unordered

Reply:
note, we first translate the dictionary into a list of tuples, then sort that list of tuples, and then encode that list of sorted tuples.